### PR TITLE
Document steps to support Docker-based HAProxy

### PR DESCRIPTION
**Pull request by fsat**
- Add installation instruction for Docker-based HAProxy.
- Add instructions for customised HAProxy reload script location.
- Add instructions for generate ConductR HAProxy configuration bundle for Docker-based HAProxy and customised HAProxy reload script location

### CODE REVIEW

File: src/main/play-doc/operation/Install.md [hunk 1]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+    option  dontlognull
+    timeout connect 5000
+    timeout client  50000
+    timeout server  50000
+
+frontend monitor
+  bind :65535
+  mode http
+  monitor-uri /test
+```
```
Comment (huntc) [inline src/main/play-doc/operation/Install.md:302]: I don't think that monitor code should be there...

File: src/main/play-doc/operation/Install.md [hunk 2]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+< HTTP/1.0 200 OK
+< Cache-Control: no-cache
+< Connection: close
+< Content-Type: text/html
+<
+<html><body><h1>200 OK</h1>
+Service ready.
+</body></html>
+* Closing connection 0
+```
```
Comment (huntc) [inline src/main/play-doc/operation/Install.md:349]: I'm unsure about this monitor endpoint - why is that needed?

**Comment by huntc** [comment]
Looking good, just a comment on the need for a monitoring endpoint. Should be necessary.

@edwardcallahan Can you please review also?

File: src/main/play-doc/operation/Install.md (hunk 2 above)
Comment (fsat) [inline src/main/play-doc/operation/Install.md:349]: Otherwise there will be no way to check if the HAProxy itself is running fine within the container.

I've done this when setting up the HAProxy within our DCOS cluster, I find it useful to have, and so I'd thought it would be useful to include it.

I'm happy to remove it though.

**Comment by fsat** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;1fGGzQTg&#x2F;312-write-doco-for-docker-based-haproxy-installation"> Write doco for Docker based HAProxy installation</a>

File: src/main/play-doc/operation/Install.md (hunk 2 above)
Comment (huntc) [inline src/main/play-doc/operation/Install.md:349]: I’d prefer to not expose endpoints unless they’re essential.

File: src/main/play-doc/operation/Install.md [hunk 3]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+
+We will use HAProxy version 1.5 official Docker image published to [Docker Hub](https://hub.docker.com/_/haproxy/).
+
+#### Prerequisite
+
+Docker needs to be installed on the host where HAProxy docker container will run.
+
+#### Preparing HAProxy config
+
+Setup basic HAProxy configuration for the Docker based HAProxy. Ensure the `conductr-agent` user which runs the ConductR-HAProxy has ownership to the HAProxy config file.
```
Comment (markusjura) [inline src/main/play-doc/operation/Install.md:278]: `Ensure that the` instead of `Ensure the`.

File: src/main/play-doc/operation/Install.md [hunk 4]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+
+```
+[172.17.0.1]$ docker run -d --name haproxy -p 10.0.7.118:9000:9000 -p 10.0.7.118:9999:9999 -p 10.0.7.118:65535:65535 -v /etc/haproxy/haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg haproxy:1.5
+```
+
+The container has `haproxy` as its name.
+
+The proxy configuration located at `/etc/haproxy/haproxy.cfg` is mounted within the container on `/usr/local/etc/haproxy/haproxy.cfg`.
+
+The container exposes the port `9000`, `9999`, and `65535` to expose default port for ConductR HTTP-based endpoints, Visualizer bundle, and HAProxy test endpoint. Additional ports for your endpoints can be exposed through the `-p` option as required.
```
Comment (markusjura) [inline src/main/play-doc/operation/Install.md:316]: I'd say: "The container exposes the ports `9000`, `9999` and `65535` to the host to make the defaults ports of the ConductR HTTP-based endpoints, the Visualizer bundle and the HAProxy test endpoint available."

Note that the port `65535` can be removed here in case you remove the `frontend monitor` configuration in the `haproxy.cfg` as well, as Chris suggested in https://github.com/typesafehub/conductr-doc/pull/251/files#r75637541.

File: src/main/play-doc/operation/Install.md [hunk 5]
```diff
@@ -276,10 +373,17 @@ We will limit the bundle's sudo privileges to running `/usr/bin/reloadHAProxy.sh
 [172.17.0.1]$ echo "conductr-agent ALL=(root) NOPASSWD: /usr/bin/reloadHAProxy.sh" | sudo tee -a /etc/sudoers
 ```
 
-On RHEL and CentOS it may also be neccessary to [disable default requiretty](https://bugzilla.redhat.com/show_bug.cgi?id=1020147) for the `conductr-agent` user in `sudoers`.
+#### Using HAProxy reload script from a customised location
```
Comment (markusjura) [inline src/main/play-doc/operation/Install.md:376]: `customized` instead of `customised` (We generally use american english in our docs)

File: src/main/play-doc/operation/Install.md [hunk 6]
```diff
@@ -276,10 +373,17 @@ We will limit the bundle's sudo privileges to running `/usr/bin/reloadHAProxy.sh
...
+
+HAProxy reload script location can be customised. ConductR-HAProxy will install its reload script in this location upon startup.
```
Comment (markusjura) [inline src/main/play-doc/operation/Install.md:378]: `customized` instead of `customised` (We generally use american english in our docs)

File: src/main/play-doc/operation/Install.md [hunk 7]
```diff
@@ -300,17 +404,182 @@ Append the `haproxy` role to the default `web` role as follows:
...
+
+Append the `/tmp/custom-haproxy-conf/runtime-config.sh` file with the following entries.
+
+```bash
+export CONDUCTR_HAPROXY_CONFIG_OVERRIDE="$CONFIG_DIR/haproxy-override.cfg"
+export HAPROXY_BIND_HOST=0.0.0.0
+export HAPROXY_RELOAD_SCRIPT_SOURCE="$CONFIG_DIR/reloadHAProxy.sh"
+```
+
+The `CONDUCTR_HAPROXY_CONFIG_OVERRIDE` and `HAPROXY_RELOAD_SCRIPT_SOURCE` supplies the customised HAProxy configuration and reload script respectively.
```
Comment (markusjura) [inline src/main/play-doc/operation/Install.md:533]: `customized` instead of `customised`.

**Comment by markusjura** [comment]
Minor comments, otherwise LGTM.

File: src/main/play-doc/operation/Install.md [hunk 8]
```diff
@@ -233,7 +233,11 @@ _Perform each step in this section on all nodes: `172.17.0.1`, `172.17.0.2` and
 
 Proxying application endpoints is required when running more than one instance of ConductR; which should be always for production style scenarios. Proxying endpoints permits connectivity from both external callers and for bundle components to communicate with other bundle components. This also allows an external caller to contact an application that is running on any ConductR node by contacting any proxy instance.
 
-We will be using `HAProxy`. Install HAProxy version 1.5 or newer.
+We will be using `HAProxy` version 1.5 or newer.
+
```
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:237]: It seems I have the choice of installing HAProxy Docker or Native. If so, why might I prefer one v the other and which is recommended.

File: src/main/play-doc/operation/Install.md [hunk 9]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+Setup basic HAProxy configuration for the Docker based HAProxy. Ensure the `conductr-agent` user which runs the ConductR-HAProxy has ownership to the HAProxy config file.
 
 ```bash
+[172.17.0.1]$ sudo mkdir /etc/haproxy
+[172.17.0.1]$ sudo touch /etc/haproxy/haproxy.cfg
 [172.17.0.1]$ sudo chown conductr-agent:conductr-agent /etc/haproxy/haproxy.cfg
 ```
 
+Populate the `/etc/haproxy/haproxy.cfg` file with the following configuration.
```
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:286]: Can we make this easier for operators to get at? Pop a copy on the downloads/conductr bucket namespace for example. 
Populate the `/etc/haproxy/haproxy.cfg` [file](https://downloads.typesafe.com/conductr/haproxy.cfg) with ...

File: src/main/play-doc/operation/Install.md (hunk 1 above)
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:302]: We certainly need to comment on the opening. What is it for? I should like to know more about this before allowing it.

File: src/main/play-doc/operation/Install.md (hunk 4 above)
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:316]: Ah, this is why I might want to 'go native'.

In light of
http://stackoverflow.com/questions/28717464/docker-expose-all-ports-or-range-of-ports-from-7000-to-8000#comment51070632_28721643
Maybe we suggest using small ranges, 9000-9050, to simplify things without so many threads to wait for

File: src/main/play-doc/operation/Install.md [hunk 10]
```diff
@@ -255,19 +259,112 @@ On some Debian distributions you may need to add a dedicated Personal Package Ar
...
+```
+
+The container has `haproxy` as its name.
+
+The proxy configuration located at `/etc/haproxy/haproxy.cfg` is mounted within the container on `/usr/local/etc/haproxy/haproxy.cfg`.
+
+The container exposes the port `9000`, `9999`, and `65535` to expose default port for ConductR HTTP-based endpoints, Visualizer bundle, and HAProxy test endpoint. Additional ports for your endpoints can be exposed through the `-p` option as required.
+
+The ports exposed by the container is bound to the `10.0.7.118` address as it's the interface where external traffic is expected to arrive from. Substitute the `10.0.7.118` address with the correct address in your environment.
+
```
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:319]: something like 
 where requests will arrive from 
or 
 the service delivery interface

'external traffic'  is not correct enough here for me

File: src/main/play-doc/operation/Install.md [hunk 11]
```diff
@@ -830,4 +1372,4 @@ NAME           HOST       USER  STATE  ID
 visualizer-1   10.0.3.75  root    R    6e68f055d1f5715ad3ff19172fa5efaf_0f6e119c-9288-425a-89e3-36379dcaccda
 ```
```
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:1374]: Should we curl an endpoint from `conduct services` or the likes?
IOW, as an operator, where's my dmarc interface? The service might be returning 200's, but if it's not reachable, it's not verified.

File: src/main/play-doc/operation/Install.md (hunk 2 above)
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:349]: Don't remove it. Move it to a troubleshooting or as part of verification (if it doesn't verify) section.

File: src/main/play-doc/operation/Install.md (hunk 8 above)
Comment (edwardcallahan) [inline src/main/play-doc/operation/Install.md:237]: Ah, with coreOS I must use Docker because I don't have apt/yum, no?

Comment (edwardcallahan) [reply src/main/play-doc/operation/Install.md:237]: Suggested text:

ConductR-HAProxy requires `HAProxy` version 1.5 or newer. This can be satisfied by installing HAProxy natively, such as using a package manager, or by running the HAProxy Docker image. Native HAProxy installation requires less configuration but may not be available such as when using CoreOS on nodes.
